mlir/include/mlir/Interfaces/AggregatedOpInterface.td

```diff
@@ -0,0 +1,57 @@
+//===- AggregatedOpInterface.td ----------------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_AGGREGATEDOPINTERFACE
+#define MLIR_AGGREGATEDOPINTERFACE
+
+include "mlir/IR/OpBase.td"
+
+def AggregatedOpInterface : OpInterface<"AggregatedOpInterface"> {
+  let description = [{
+    This Interface is particularly useful in cases where we have an operation
+    that can be lowered into a sequence of simpler operations, thus essentially
+    decomposing an operation into a set of one or many simpler operations.
+    The operation being decomposed need to implement this Interface by implementing
+    the method `decomposeOperation` and return the set of values which would replace
+    the uses of the operation being decomposed.
+    Eg:
+        Assume there is an operation `CustomOp_Mul_Add` that takes in an input tensor
+        and a constant. It basically performs element-wise multiplication of the input
+        tensor with the given constant, and then performs element-wise addition of the
+        intermediate resulting tensor with the given constant.
+        `CustomOp_Mul_Add` can thus essentially be decomposed by implementing this
+        Interface.
+        `linalg::SoftmaxOp` is one such operation which makes use of this Interface
+        for implementing its decomposition.
+  }];
```

Comment:
This seems ill-defined to me: the whole concept of MLIR and progressive lowering is that almost **every** operation will be decomposed, gradually, through many layer of abstractions: so most ops (above LLVM) are fitting this description.

I still don't know when I would use this interface and ask to "decompose" such an operation, what is the use-case?



(also the `Aggregate` naming confused me because `Aggregate` is how LLVM named "struct" types).

Reply:
> also the Aggregate naming confused me because Aggregate is how LLVM named "struct" types).

Oh okay, I understand. We can change the naming here perhaps, in that case.

> I still don't know when I would use this interface and ask to "decompose" such an operation, what is the use-case?

I understand your point. I'll try to explain the use-case which necessitates this NFC change.
So, the idea is to perform [a similar operation like linalg::Softmax](https://github.com/llvm/llvm-project/blob/85f6b2fac9a367337e43ca288c45ea783981cc16/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2550) for a few other [Linalg-Like operations in IREE](https://github.com/openxla/iree/pull/15294) by implementing [the already defined upstream interface](https://github.com/llvm/llvm-project/blob/85f6b2fac9a367337e43ca288c45ea783981cc16/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L820) used by [linalg::Softmax here on top-of-main](https://github.com/llvm/llvm-project/blob/85f6b2fac9a367337e43ca288c45ea783981cc16/mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td#L97) and then instead of having separate individual passes for their decomposition, we only have one pass which would do the needful.

I'm not sure if we're discussing about removing this interface altogether from top-of-main now or we shouldn't move it from `linalg::` namespace to a more generic namespace - since the Interface already exists.

Can you confirm once? @joker-eph 

Also, perhaps @MaheshRavishankar can add a better context here if I missed any pointers pertaining to the use-case.

Reply:
For context, when this interface was first introduced the concept we wanted to capture is instructions that can be represented with a sequence of simpler instructions within the same dialect.

Reply:
The same thing could be achieved with providing proper pattern rewriters, but we found that it was easier to discover when these particular patterns were part of the related "instruction" itself.

At the time we kept this interface within linalg because we would have expected that using that more broadly would warrant a proper RFC.

Reply:
See https://reviews.llvm.org/D154363 for more context

Reply:
I have a hard time seeing how this is something that composes across dialects or can just be reused generically. That seems quite "specific" to a specific pass at one given level of abstraction (and this is what makes it pretty contextual and not generic as I see it right now).

Reply:
Echo-ing @qcolombet comments. All operations eventually get decomposed, you are right. As Quentin mentioned you could have a pattern rewrite that decomposes an operation into smaller parts. We might maybe need a better term for it, but  this interface is meant to group together a set of operations that exhibit a common behavior. Some examples are Softmax (that is in tree), FlashAttention and few others (that are actually out-of-tree in IREE). These are all "concepts" that are common in typical ML programmer land, and maintaining these as an "aggregate operation" in the compiler on the input better represents the programmer intent. You could do several transformations on these aggregate operations themselves (like tiling), but at some point in the compilation flow, it is better to lower them into the decomposed sequence of operations (typically a DAG of linalg-like operations) so that the rest of the compilation flow works on the decomposed sequence (like reductions or matmuls, etc).  All these operations seem to follow a similar compilation flow, and the goal is to have some interfaces that these operations can implement so that the compiler just works on these interfaces. There is an in-tree example for now, softmax. We can implement the same interface on FlashAttention etc. that are not in tree as of now. (We plan to move those in tree as well if that helps, just like Softmax was moved into tree by Quentin)

So if you have a better name we could use that. (Naming is hard :) ). Also note that this is just moving an interface that is already in tree, but within Linalg, out of Linalg dialect. There doesnt seem to be a reason for this interface to live in Linalg dialect.

Reply:
> Also note that this is just moving an interface that is already in tree, but within Linalg, out of Linalg dialect. There doesnt seem to be a reason for this interface to live in Linalg dialect.

I assume our messages crossed and I expanded on this above?

Reply:
> I have a hard time seeing how this is something that composes across dialects or can just be reused generically. That seems quite "specific" to a specific pass at one given level of abstraction (and this is what makes it pretty contextual and not generic as I see it right now), and that's why it seems fine to me where it is.
> 
> For example would I implement this interface for the Arith dialect by splitting my `fma` into `add` + `mul`? This may be something I want for my use-case, but if I do this it'll be automatically picked up by the transformation when you apply it at the linalg level.

Yeah thats not the intent, but there is no way to codify it AFAICS. It is hard to predict where all these gets used. If there are issues with this, I am happy to go with one of two options
1) We keep the interface in Linalg...
2) We move the interface out of core into IREE...  Since there we have more control of which ops implement the interface, we can prevent unintended uses of it.

Reply:
I'm with these 2 options, but none of them is consistent with "land this PR" right now.

Reply:
> For example would I implement this interface for the Arith dialect by splitting my `fma` into `add` + `mul`? This may be something I want for my use-case, but if I do this it'll be automatically picked up by the transformation when you apply it at the linalg level.

That's a good point and I believe we wouldn't want that thing to be picked up at the linalg level. That being said, this is transformation specific, meaning we could have the transformation pick the level it wants or all levels if it is what the specific use case is. E.g., `if op.implementThisInterface and op.isDialect(linalg)`. I'm not saying this is what we want (nor if it is even possible), but assuming it is, then that could be a way to have the interface being more controllable.

Not arguing either way for this PR (accept or reject), ultimately what I wanted to achieve with the original PR (admittedly without giving it much thoughts) was to do a better job at describing / discovering what is supported by the conversion patterns and what should be expanded before conversion. The idea of this interface was to capture what needed to be expanded.